### PR TITLE
Era based effective area to be used in Muon selectors

### PR DESCRIPTION
**Pull request by bmahakud**
#### PR description:
- This PR is made after closing the PR #26292
- Adding Era dependent effective area for mini isolation calculation
- The ticket that concerns this is : https://its.cern.ch/jira/browse/CMSMUONS-206
- The effective areas that are used to calculate the mini Isolation changes with Era
- All comments from the previous PR are addressed here. 
- @perrotta @slava77 @folguera @drkovalskyi 
Please look at the Era implementation. This is important as newly created tags run2_muon_2016, run2_muon_2017 and run2_muon_2017 will be used in other future muon POG related PRs.
--

### CODE REVIEW

File: Configuration/Eras/python/Era_Run2_2016_cff.py [hunk 1]
```diff
@@ -5,6 +5,9 @@
 from Configuration.Eras.Modifier_stage2L1Trigger_cff import stage2L1Trigger
 from Configuration.Eras.Modifier_ctpps_2016_cff import ctpps_2016
 from Configuration.Eras.Modifier_run2_HLTconditions_2016_cff import run2_HLTconditions_2016
+from Configuration.Eras.Modifier_run2_muon_2016_cff import run2_muon_2016
 
-Run2_2016 = cms.ModifierChain(run2_common, run2_25ns_specific, stage2L1Trigger, ctpps_2016, run2_HLTconditions_2016)
+
+
+Run2_2016 = cms.ModifierChain(run2_common, run2_25ns_specific, stage2L1Trigger, ctpps_2016, run2_HLTconditions_2016, run2_muon_2016)
```
Comment (slava77) [inline Configuration/Eras/python/Era_Run2_2016_cff.py]: here and in other files please add a linebreak for readability and drop the extra newline(s) before "Run2_2016 = "

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py [hunk 2]
```diff
@@ -30,6 +30,12 @@ def miniAOD_customizeCommon(process):
     process.patMuons.computeSoftMuonMVA = True
 
     process.patMuons.addTriggerMatching = True
+    from Configuration.Eras.Modifier_run2_muon_2016_cff import run2_muon_2016
+    from Configuration.Eras.Modifier_run2_muon_2017_cff import run2_muon_2017
+    from Configuration.Eras.Modifier_run2_muon_2018_cff import run2_muon_2018
+    run2_muon_2018.toModify( process.patMuons, effectiveAreaVec = cms.vdouble(0.0566, 0.0562, 0.0363, 0.0119, 0.0064))
+    run2_muon_2017.toModify( process.patMuons, effectiveAreaVec = cms.vdouble(0.0566, 0.0562, 0.0363, 0.0119, 0.0064))
+    run2_muon_2016.toModify( process.patMuons, effectiveAreaVec = cms.vdouble(0.0735,0.0619,0.0465,0.0433,0.0577))
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

so ```run2_muon_2016.toModify( process.patMuons, effectiveAreaVec = [0.0735,0.0619,0.0465,0.0433,0.0577]```.
Perhaps also keeping 16 -> 17 -> 18 order is better

File: PhysicsTools/PatUtils/src/MiniIsolation.cc [hunk 3]
```diff
@@ -55,19 +55,21 @@ PFIsolation getMiniPFIsolation(const pat::PackedCandidateCollection *pfcands,
...
 
-  float muonRelMiniIsoPUCorrected(const PFIsolation& iso,
-				  const math::XYZTLorentzVector& p4,
-				  float dr,
-				  float rho)
+  double muonRelMiniIsoPUCorrected(const PFIsolation& iso,
+				   const math::XYZTLorentzVector& p4,
+				   double dr,
+				   double rho,
+                                   std::vector<double> &EA)
```
Comment (slava77) [inline PhysicsTools/PatUtils/src/MiniIsolation.cc]: start local variables and data members with lower case letter. Capitalization is reserved for class names.
Also, keep the vector const.
==> ```const std::vector<double> &ea```

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
there are no changes in the selector values.
It seems like something is not working.

**Comment by bmahakud** [comment]
@slava77  I will try check soon to know the expected changes.

**Comment by slava77** [comment]
@bmahakud 
please provide a link to slides which describe the new EA documentation.
Thank you.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/26489	07e6175
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences
    - local test with wf  136.8311 1K events  (x10 of what's in jenkins) showed some differences. In fact there is a diff even in the cms-bot output, but it requires a manual comparison to see a change, which is not visible at the histogram bin content level.

**Comment by santocch** [comment]
+1

**Comment by bmahakud** [comment]
@slava77 
- I did a event by event comparison  of print report of miniIsolation while running runTheMatrix -l 10824.0 (TTbar_13TeV_TuneCUETP8M1) and found nothing suspicious.
- From code point of view everything seems to work fine. 
- I did print 1.) Effective area, 2.) miniIsolation value, 3.) correction to isolation (depends on effective area) and 4.) muon Eta at this point https://github.com/bmahakud/cmssw/blob/EraBasedEffectiveAreaFinalPR/PhysicsTools/PatUtils/src/MiniIsolation.cc#L72
- The print report  (2018 vs 2016)shows infact code works fine but changes in miniIsolation values look very small, that could be the reason why you saw negligible differences 
- Print report available at : https://github.com/bmahakud/bibhuRepo/blob/master/PrintReportWith2016EffectiveArea.txt
https://github.com/bmahakud/bibhuRepo/blob/master/PrintReportWith2018EffectiveArea.txt

**Comment by bmahakud** [comment]
> @bmahakud
> please provide a link to slides which describe the new EA documentation.
> Thank you.

@folguera   Can you help in getting links of slides as Slava has asked.

**Comment by slava77** [comment]
@fabiocos @kpedro88 
another reco/miniAOD PR relies on this one.
please check this PR and merge or propose modifications at your earliest.
Thank you.

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
merge

**Comment by fabiocos** [comment]
+operations

the addition of eras is coherent with the purpose of the PR

**Pull request merged**